AG-11673 Create area-series level stylers

olegat opened the pull request:
https://ag-grid.atlassian.net/browse/AG-11673

olegat committed:
AG-11673 Add option `series[type='area'].styler`

olegat committed:
AG-11673 Fix lineSeries.test.ts compilation

olegat committed:
AG-11673 Add initial test for area series

olegat committed:
AG-11673 Make areaSeries.ts/lineSeries.ts easier to diff

olegat committed:
AG-11673 Implement `series[type='area'].styler`

olegat committed:
AG-11673 Use default `marker.enabled: true` if styler returns marker styling

olegat committed:
AG-11673 Add test for deriveMarkerEnabledFromStyler

olegat committed:
AG-11673 Fix `strokeWidth` $isUserOption expansion using stylers

When the `styler` callback returns a `stroke` color, we want the `strokeWidth`
value to default to 2 instead of 0.

olegat committed:
AG-11673 Remove AreaSeries.deriveMarkerEnabledFromStyler

olegat committed:
AG-11673 Fix autoEnabled for "userPartial" edgeValues

olegat committed:
AG-11673 nx format

olegat committed:
AG-11673 Update deriveMarkerEnabledFromStyler snapshot

lsjroberts commented on packages/ag-charts-community/src/chart/optionsGraphService.ts:
```diff
@@ -1,15 +1,25 @@
+
+interface Graph {
+    resolvePartial(path: Array<string>, partialOptions?: PlainObject, opts?: ResolvePartialOpts): Resolved;
+    hasUserOption(path: Array<string>): boolean;
```
As discussed - let's remove this option as its risky to expose and hopefully not needed with our bug fixes.

olegat replied:
Done. `hasUserOption` has been removed. Instead, `AreaSeries.updateDatumSelection` resolves the `marker.enabled` property using the graph service https://github.com/ag-grid/ag-charts/commit/d005bbf86d8fccb45f3a1ce2330a26e7c6f15e7f

lsjroberts commented on packages/ag-charts-community/src/module/optionsGraph.ts:
```diff
@@ -246,7 +246,7 @@ export class OptionsGraph extends AdjacencyListGraph<unknown, string> implements
      * Returns an object with only those keys that were also present within `partialOptions`.
      */
-    resolvePartial(path: Array<string>, partialOptions?: PlainObject) {
+    resolvePartial(path: Array<string>, partialOptions?: PlainObject, opts?: { pick?: boolean }) {
```
`pick` option makes sense. For context, this solves the issue with `strokeWidth` being dependent on `stroke` and needing to get back all the changed values in some uses. Compared to the other cases which wanted to not override the existing values.

lsjroberts commented on packages/ag-charts-community/src/module/optionsGraph.ts:
```diff
@@ -305,7 +312,10 @@ export class OptionsGraph extends AdjacencyListGraph<unknown, string> implements
+            return (
+                this.findNeighbour(pathVertex, USER_OPTIONS_EDGE) != null ||
+                this.findNeighbour(pathVertex, USER_PARTIAL_OPTIONS_EDGE) != null
+            );
```
As discussed - this bug should be fixed by the similar changes made here https://github.com/ag-grid/ag-charts/pull/4959/files#diff-27b8479ea6fa70d9276fa0fb48c97885b34220e6c9169f5b4fafb808d98b4757R414-R417

olegat replied:
This change was reverted in favour of the existing fix in the other PR

lsjroberts reviewed (approved):
OptionsGraph changes LGTM with changes as discussed.

olegat committed:
Merge remote-tracking branch 'latest' into AG-11673/area

Conflicts:
	packages/ag-charts-community/src/api/agCharts.ts
	packages/ag-charts-community/src/chart/optionsGraphService.ts
	packages/ag-charts-community/src/module/optionsGraph.ts

olegat committed:
AG-11673 Fix compilation after conflict resolution

olegat committed:
AG-11673 nx format

olegat committed:
AG-11673 Remove `OptionsGraphService.hasUserOption`

olegat committed:
AG-11673 Remove obsolete commented code

olegat committed:
AG-11673 Fix marker.enabled resolution in area series

olegat committed:
AG-11673 Fix test snapshots

olegat committed:
AG-11673 Fix tests

FAIL
 ag-charts-community - unit  packages/ag-charts-community/src/api/preset/sparkline.test.ts
   Sparkline Preset › #updateDelta › should update a bar sparkline with an item styler

 ag-charts-community - e2e  packages/ag-charts-community/src/chart/callbacks.test.ts
   callback cache › AG-10112 re-evaluate callbacks on update

WARN (Console)
 ag-charts-enterprise - e2e  packages/ag-charts-enterprise/src/series/range-bar/rangeBar.test.ts
   RangeBarSeries › AG-15448 › should render updated data in the itemStyler specified colors
